src/bindings/jna/libelektra/src/main/java/org/libelektra/KeySet.java

```diff
@@ -263,6 +259,22 @@ public int size ()
 		return new KeySet (Elektra.INSTANCE.ksCut (getPointer (), cutpoint.getPointer ()));
 	}
 
+	/**
+	 * Removes the the specified key from key set
+	 *
+	 * @param key Key to remove
+	 * @return True, if the key was found and removed, false otherwise
+	 * @throws KeySetReleasedException  if this {@link KeySet} has already been
+	 *                                  released
+	 * @throws KeyReleasedException     if {@code key} has already been released
+	 * @throws IllegalArgumentException if {@code key} is {@code null}
+	 */
+	@Nonnull public boolean remove (Key key)
```

Comment:
LGTM, except for two things:

1. We should return `Key` here. The `Key` returned by lookup is (normally) not the same as the `Key` you pass in. Only the name will be the same. This would also match the signature of the standard `Map.remove` method.
2.  I would add a `public Key remove (String name)` version that uses `ksLookupByName` for convenience.

Reply:
> 1. We should return `Key` here. The `Key` returned by lookup is (normally) not the same as the `Key` you pass in. Only the name will be the same. This would also match the signature of the standard `Map.remove` method.
The signature of `Map.remove` is slightly different, in taking `Object` instead of `Key` as well as returning null instead of throwing an exception. As long as we do not implement the Map interface, we should not care that much about mimicking its contracts.
I will change the return type and introduce throwing a NoSuchElementException

Reply:
> introduce throwing a NoSuchElementException

I wouldn't do that. I would just return `null`, if nothing was removed.

Reply:
done

Reply:
> > introduce throwing a NoSuchElementException
> 
> I wouldn't do that. I would just return `null`, if nothing was removed. Then a caller can get back the same `boolean` by just doing the `!= null` outside the function.

I don't like returning null values... but it comes down to the following question:
What's the use-case for calling remove when i do not know whether the key is contained within the key set?

Reply:
btw. `KeySet::remove(int)` currently also follows a "removes specified key..." semantic. if we go for the "tries..." approach, we may have to update this method...

Reply:
> I don't like returning null values...

Yes, `null` values are annoying to deal with, **but** using exceptions as return values (for control flow) is _much much_ worse IMO.

The semantics of `remove(Key)` are well defined, if the element doesn't exist. The goal here is simply that the given `Key` is not present in the `KeySet` after calling `remove`. Therefore we can just do nothing if the key is not present, throwing a `NoSuchElementException` is not a good solution IMO.

For `remove(int)` things are different. Here we want to explicitly remove the element at index X. The goal is _not_ that there is no key at index X. For that we would have to remove every key from X to the end of the keyset. So for an out-of-bounds index the behaviour is not well-defined.

> i would consider it better and more consistent to return an Optional.

Yes, returning an `Optional<Key>` is also an option. I agree that this would also be more consistent with the rest of the API.

Reply:
@kodebach I agree, thank you for taking the time for such thorough feedback! done.